cpp/oneapi/dal/backend/primitives/reducer.cpp

```diff
@@ -0,0 +1,221 @@
+            el                       = vectors[group_id * global_dim + i * local_dim];
+            partial_reduces[item_id] = binary(partial_reduces[item_id], unary(el));
+        }
+
+        idx.barrier(cl::sycl::access::fence_space::local_space);
+
+        for (std::uint32_t stride = local_size / 2; stride > 1; stride /= 2) {
+            if (stride > item_id) {
+                partial_reduces[item_id] =
+                    binary(partial_reduces[item_id], partial_reduces[item_id + stride]);
+            }
+
+            idx.barrier(cl::sycl::access::fence_space::local_space);
+        }
+
+        if (item_id == 0) {
+            reduces[group_id] = binary(partial_reduces[item_id], partial_reduces[item_id + 1]);
+        }
+    }
+
+private:
+    const std::uint32_t vector_size, n_vectors;
+    const Float* const vectors;
+    Float* const reduces;
+    local_acc_t partial_reduces;
+};
+
+} // namespace impl
+
+template <typename Float, typename BinaryFunctor, typename UnaryFunctor, layout Layout>
+reducer_singlepass<Float, BinaryFunctor, UnaryFunctor, Layout>::reducer_singlepass(
+    cl::sycl::queue& queue,
+    BinaryFunctor binary_func,
+    UnaryFunctor unary_func)
+        : binary(binary_func),
+          unary(unary_func),
+          q(queue),
+          max_work_group_size(std::min<std::int64_t>(
+              queue.get_device().template get_info<cl::sycl::info::device::max_work_group_size>(),
+              256)) {}
+
+template <typename Float, typename BinaryFunctor, typename UnaryFunctor, layout Layout>
+cl::sycl::event reducer_singlepass<Float, BinaryFunctor, UnaryFunctor, Layout>::operator()(
+    const Float* input,
+    Float* output,
+    std::int64_t vector_size,
+    std::int64_t n_vectors,
+    std::int64_t work_items_per_group) {
+    const std::int64_t local_buff_size = std::min(this->max_work_group_size, work_items_per_group);
+    auto event                         = this->q.submit([&](cl::sycl::handler& handler) {
+        typedef cl::sycl::
+            accessor<Float, 1, cl::sycl::access::mode::read_write, cl::sycl::access::target::local>
+                local_acc_t;
+        auto partial_reduces = local_acc_t(cl::sycl::range<1>{ local_buff_size }, handler);
+        auto functor_instance =
+            kernel_t{ /*.binary          =*/binary,
+                      /*.unary           =*/unary,
+                      /*.vector_size     =*/static_cast<std::uint32_t>(vector_size),
+                      /*.n_vectors       =*/static_cast<std::uint32_t>(n_vectors),
+                      /*.vectors         =*/input,
+                      /*.reduces         =*/output,
+                      /*.partial_reduces =*/partial_reduces };
+        const cl::sycl::range<2> local_range{ static_cast<size_t>(work_items_per_group), 1 };
+        const cl::sycl::range<2> global_range{ static_cast<size_t>(work_items_per_group),
+                                               static_cast<size_t>(n_vectors) };
```

Comment:
Casting to `std::size_t` is not safe. There is a special function that safely creates ranges, see https://github.com/oneapi-src/oneDAL/blob/3709dcc719974c93f30b3ef25da815b7082b6982/cpp/oneapi/dal/backend/common.hpp#L67

Similar function can be written for 2d ranges.